### ISSUE TITLE
Proposal: Highly Performant Scrolling Experiences

### ISSUE THREAD

**Issue by tannermeade** [superlistapp/super_editor#283]
This proposal outlines some of the challenges associated with delivering a highly performant scrolling experience built on best practices as advised by the core Flutter team.

# The Current Problem

SuperEditor uses a `SingleChildScrollView` built in `_DocumentInteractorState._buildDocumentContainer()` to display all the content built in `DefaultDocumentLayout.build(BuildContext context)`.

The `SingleChildScrollView` only uses the `RenderBox protocol` to create the scrolling experience which “doesn’t work great for scrolling” (evidences below).

The layout protocol for scrolling in Flutter is the `Sliver protocol` available in the [RenderSliver object](https://api.flutter.dev/flutter/rendering/RenderSliver-class.html).

Problems:

- Scrolling experience degrades as the document gets larger.
- Scrolling can become unstable and unusable.
- Responsiveness for clicking and highlighting suffers a huge performance hit.
- Gets bad enough that performance makes double-clicking a word impossible.

# The Best Practice

On the topic of scrolling experiences Ian (Flutter Tech Led) said in the [“Sliver Explained” Boring Show episode](https://youtu.be/Mz3kHQxBjGg?t=304):

> Pretty much everything that scrolls in Flutter is using Slivers.
> 
> There's one exception. We have the [SingleChildScrollView]. That is a viewport that lets you scroll a single item around. That's the exception, and that one you don't usually use, because it's not lazy.
> 
> It's only useful if … you're worried that if the screen is too small, or if the keyboard comes up and shrinks the dialog, then it might not quite fit, and so you'll want to scroll inside that.
> 
> So there, it doesn't matter if you're lazy, because you're going to have everything instantiated anyway normally.
> 
> But anytime you have a list, where you don't know how many things are in the list -- you use ListView, you use GridView, CustomScrollView, all of those that you think Slivers under the hood.

# Evidence

I personally ran into this problem while making an app with SuperEditor. It is considered a blocking issue for us. Below I've tried to extract the problem to a demonstratable situation. The document involved is quite large, however, even worse performance can be found with much smaller documents.

Flutter 2.2.3 • channel stable
SuperEditor `main` Branch, latest commit 27ecb1a122b27aeb4f0f523490108ab4622e3dcd
Displaying the same `MutableDocument`
`web` as target

## Main branch performance in debug and release modes:

https://user-images.githubusercontent.com/46303912/131229218-69deb9d1-da54-4c3a-a57a-876e14f9a7c7.mov

Scrolling in debug takes at least half second to respond and is very rough, so I didn't scroll too much in the video.

## Experimental Sliver Implementation

Here is my experimental implementation branch: https://github.com/tannermeade/super_editor/tree/sliverLayout

Experimental Sliver implementation performance in debug and release modes:

https://user-images.githubusercontent.com/46303912/131228980-9b99cbb0-46ca-461d-bd5f-9b93cc5c0ade.mov

**Comment by tannermeade** [comment]
Some of the decisions to be made here surround questions like:

- Should SuperEditor use the Sliver protocol as the default layout method?
- Should the `SingleChildScrollView` be used at all?
- How can both `SingleChildScrollView` and the sliver protocol be supported?

There is a coupling problem between the `DocumentInterator` and the `DocumentLayout` that will need some decisions if SuperEditor is going to support any Layout protocol because the `DocumentInterator` controls the scrolling layout where `DocumentLayout` probably should.

In my experimental branch (sliverLayout) I played around with using a CustomScrollView in the `DocumentInterator`. I believe a solution that would allow the `DocumentLayout` object to provide a `SliverToBoxAdapter` if it wants to use similar functionality of a `SingleChildScrollView` is a good and valid approach.

**Comment by tannermeade** [comment]
This issue is mainly to get the SuperList team's feedback as I turn my experimental branch into a pull request.

**Comment by wilsonowilson** [comment]
Even though using `SingleChildScrollView` isn't the most performant approach, I find it to be the most reasonable; the reason being that it allows you to get important information about the position and location of built `DocumentNode`s. Since `ListView`s don't render any off-screen items, getting information about the widget becomes extremely difficult if not impossible.

The size and position of the built nodes are especially important for features like ensuring visibility of off-screen nodes when typed, possible find and replace et cetera.

That being said, there are a few things that can be done to improve super_editor's performance. 

For starters, I believe the reason why the selection is so unperformant is that the whole `DocumentLayout` gets rebuilt when the composer's selection changes. I managed to fix this in my own custom editor built on super_editor by moving the listener to the block level and only rebuilding the block when the block selection enters, leaves or changes.

_The following examples use debug mode with ~50,000 words, 1375 blocks, and 13940-pixel high SingleChildScrollView [My custom editor]_

### Before moving the selection listener to the block-level

https://user-images.githubusercontent.com/63465656/131259840-d131d960-645c-4d96-a5aa-ad78b4608aec.mov

### After moving the selection listener to the block-level

https://user-images.githubusercontent.com/63465656/131260036-ff218cc3-30c9-4ccc-bc92-835bfbe7aa0a.mov

I'm not so sure what to do about scrolling, but on my editor which also uses a `SingleChildScrollView` and `Column` for its layout, I'm getting an almost perfectly consistent 60fps on release with over 2000 blocks and 60,000 words.

That being said, I came up with a hacky workaround for the column's scrolling performance problems using `Visibility` and the `ScrollPosition` which actually produces minor, but visible results. 

By making off-screen widgets "invisible", I was able to increase the scrolling fps for a 27,000-word document from around 45fps to a solid 60fps.


### Before making non-visible widgets invisible

https://user-images.githubusercontent.com/63465656/131261142-b8f2a365-e8fe-4c56-ab13-a09d3cadb104.mov

### After making non-visible widgets invisible

https://user-images.githubusercontent.com/63465656/131261165-b5182d3f-eb8e-44f6-aae6-ac2ed1e40363.mov


```dart
class _WidgetOverscrollMask extends StatefulWidget {
  const _WidgetOverscrollMask({
    required Key key,
    required this.child,
  }) : super(key: key);

  final Widget child;

  @override
  __WidgetOverscrollMaskState createState() => __WidgetOverscrollMaskState();
}

class __WidgetOverscrollMaskState extends State<_WidgetOverscrollMask> {
  bool _visible = false;
  late ScrollableState scrollable;
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    scrollable = Scrollable.of(context)!..position.addListener(_onScroll);
    // set current on-screen items to visible 
    WidgetsBinding.instance?.addPostFrameCallback((timeStamp) {
      _onScroll();
    });
  }

  @override
  void dispose() {
    scrollable.position.removeListener(_onScroll);
    super.dispose();
  }

  void _onScroll() {
    if (!mounted) {
      return;
    }

    final widgetBox = context.findRenderObject() as RenderBox;
    final size = widgetBox.size;
    final position = widgetBox.localToGlobal(Offset.zero);
    final scrollDistance = scrollable.position.pixels;
    final relativePosition = position.dy - scrollDistance;

    final extent = position.dy + size.height;
    // Extra padding for possible widget size changes.
    final padding = scrollable.context.size!.height * 0.4;

    if (extent > -padding &&
        relativePosition <
            (scrollable.context.size!.height - scrollDistance + padding)) {
      if (!_visible) {
        setState(() {
          _visible = true;
        });
      }
    } else {
      if (_visible) {
        setState(() {
          _visible = false;
        });
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return RepaintBoundary(
      child: Visibility(
        maintainSize: true,
        maintainState: true,
        maintainAnimation: true,
        maintainInteractivity: false,
        maintainSemantics: true,
        visible: _visible,
        child: widget.child,
      ),
    );
  }
}
```

It goes without saying that for most normal documents < 15,000 words, the `SingleChildScrollView` should be sufficient.

**Comment by tannermeade** [comment]
@wilsonowilson I'll respond properly later, but scrolling is most affected on web. All your examples are using macOS as the target.

> The size and position of the built nodes are especially important for features like ensuring visibility of off-screen nodes when typed, possible find and replace et cetera.

Can you please help me understand which use cases can not be accomplished with the Sliver protocol? Find and replace can be done with the Sliver protocol. In what use cases would you need to perform layout for a component that is 100% off screen? 🤔

**Comment by matthew-carroll** [comment]
Thanks for the proposal.

**TL;DR: For performance issues, please file tickets with specific examples that you believe should work better than they do. We can work on problems individually and then accumulate evidence for a broader change.**

One thing that I can say for sure is that the first place to start is with the problem, rather than the solution. If we have scrolling performance issues, as it appears we do, then the first step is to locate every last performance improvement that we can find in the existing abstractions. Only after making every reasonable attempt to resolve the issue in-place, should we consider a significant change in structure.

**The provided example**
One thing to note before I go into Sliver details: the first example provided appears to contain a single, massive paragraph. If that's true, that's a problem no matter what. The performance value of Slivers, much like lazy lists, is that only the visible widgets are built. Well, if you have a paragraph with 1,000 lines of text, that entire `Text` widget is going to be built and painted no matter what you do. There's nothing to lazily load or recycle because everything is one widget.

When putting together hypothetical examples, please make sure that they reflect a reasonable document structure.

**Slivers**
Slivers are absurdly complicated for what they are. When creating a custom `RenderSliver`, you're faced with something like a dozen different [dimensions of input](https://api.flutter.dev/flutter/rendering/SliverConstraints-class.html) to determine your outgoing geometry. Even the [definition of geometry](https://api.flutter.dev/flutter/rendering/SliverGeometry-class.html) is complicated. We want super_editor to be usable and extensible for average developers, but average developers don't understand slivers. I was on the Flutter team, I had to write a few `RenderSlivers`, and I still don't understand them.

It's stated that `RenderBox`s "don't work great for scrolling" and that Slivers are Flutter's "protocol for scrolling". I think that both of these statements are misleading. 

Boxes work great for scrolling, so long as you don't do things that boxes aren't great at doing. If you want a collapsing app bar, then you're right, boxes aren't great. If you want just about anything else, boxes work pretty well. 

Slivers weren't merely designed for scrolling. They were designed for relatively advanced, mobile scrolling use-cases. Do you want a collapsing, pinned app bar? Use a Sliver. Do you want an infinite list that scrolls in both directions? Use Slivers. Do you want an item in a list that might dynamically alter the size of its content as you scroll? Use a Sliver. But a rich text editor doesn't do any of these things. And Flutter had absolutely no concern for rich text editors when it was trying to reach adoption among Android and iOS phones.

_Quick caveat, all scrolling uses Slivers in Flutter. `SingleChildScrollView` is no exception. But I'm talking about what kinds of widgets are used for content, regardless of the surrounding `ScrollView`._

**What Slivers make difficult/impossible**
super_editor makes ample use of coordinate mapping. We're constantly moving from the app, to the document, to a paragraph, back to the document, back to the app. This is an (x,y) cartesian mapping up and down the render tree. Here's a related `RenderBox` method:

https://api.flutter.dev/flutter/rendering/RenderBox/globalToLocal.html

You will notice a lack of any such coordinate mapping in `RenderSliver`, because a Sliver is not a cartesian coordinate system. There is no (0, 0), and there is no static "size".

https://api.flutter.dev/flutter/rendering/RenderSliver-class.html

You might try to use the surrounding `RenderViewport` to infer the position of a Sliver, but that gets into some very hairy territory where lots of things could go wrong in that calculation.

We also run into difficulties with finding things that are offscreen. Consider that right now, you can find the exact (x,y) origin for any piece of content in the document, and you can obtain its width and height. These pieces of information are available immediately, for anything. If you use Slivers, or even a lazy list, that's not true. To find something in the document offscreen, you'll have to build every offscreen widget until you get there. If you look for something that doesn't exist, you'll end up building everything down to the bottom of the document, because you don't know if it exists until you look everywhere. So there are definitely trade-offs here.

**Performance**
Slivers and boxes aside, we need super_editor to be performant. There's no question about that. But we shouldn't ignore the trade-offs we make when switching to significantly different approaches. To avoid any unfortunate trade-offs, we should begin by solving every problem that we can without making any structural modifications. Then, we should try to get away with the least invasive modifications possible.

**Comment by tannermeade** [comment]
@matthew-carroll Thanks for the reply. I'll get to thoroughly responding to your comment by:

- Providing significant evidence that the problem is structural.
- That we don't need to derive the infinite maximum of our current structure to know there is a significant and core structure problem that requires we change structures.
- Thoroughly demonstrate that this is NOT a hypothetical problem by providing real-world examples that are within the scope of super_editor.
- Thoroughly demonstrate that RenderBoxes do NOT work pretty well for everything within the scope of super_editor.
- How RenderBox.globalToLocal() is not a requirement for super_editor where it is currently being used.
- Precisely defining the scope that super_editor actually uses coordinate mapping for.
- Defining the faults of the current use of coordinate mapping.
- Defining the faults with the CURRENT structure that already makes use of traversing the whole document.

But right now you are working on mobile support which is more important to me than the immediate progress of official scrolling performance improvements in super_editor. It would also be helpful to include mobile use cases into this discussion after they've matured a bit further.

My main purpose here was to see what needs to happen to progress the official scrolling performance improvements. I'll get to properly replying back to you in a while. 👍

**Comment by matthew-carroll** [comment]
I don't think those responses are going to be a good use of your time. As I mentioned, I don't think it makes sense to abandon cartesian coordinates before taking every other step to eliminate performance problems first. If you or @wilsonowilson see local, unnecessary performance costs, then please file a ticket per issue.

**Comment by bk-one** [comment]
@tannermeade closing this, please re-open if you still experience performance issues - or file more specific tickets.
